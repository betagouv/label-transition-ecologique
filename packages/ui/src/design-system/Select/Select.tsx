import { SelectBase, SelectProps } from './components/SelectBase';

/**
 * SÃ©lecteur de valeur unique
 *
 * Ajouter `isSearcheable`, `createProps` ou `onSearch` pour faire un Searchable select
 */
export const Select = ({ values, onChange, ...props }: SelectProps) => {
  return (
    <SelectBase
      {...props}
      //values sera toujours un OptionValue simple sans tableau
      onChange={(v) => onChange(v !== values ? v : '')}
      values={values}
    />
  );
};
