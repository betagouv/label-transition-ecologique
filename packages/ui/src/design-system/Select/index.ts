export * from './Select';
export * from './SelectMultiple';
export * from './SelectFilter';
