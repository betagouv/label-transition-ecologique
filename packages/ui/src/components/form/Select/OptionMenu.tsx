import {Ref, forwardRef} from 'react';
import DropdownFloater from '../../floating-ui/DropdownFloater';

import {CreateOption} from './Select';
import {Option} from './Options';

type Props = {
  option: Option;
  createProps: CreateOption;
};

export const OptionMenu = ({option, createProps}: Props) => (
  <DropdownFloater
    placement="top"
    offsetValue={{mainAxis: 8}}
    noDropdownStyles
    render={() => (
      <div
        className="flex flex-col divide-y divide-x-0 divide-solid divide-grey-3 rounded overflow-hidden bg-white border border-grey-3"
        onClick={evt => {
          evt.stopPropagation();
        }}
      >
        <button
          className="flex items-center w-full py-2 pr-4 pl-3 text-sm text-grey-8"
          onClick={() => createProps.onUpdate(option.value, 'TODO')}
        >
          <span className="flex fr-icon-edit-line before:w-4 before:h-4 mr-2" />
          Éditer
        </button>
        <button
          className="flex items-center w-full py-2 pr-4 pl-3 text-sm text-grey-8"
          onClick={() => createProps.onDelete(option.value)}
        >
          <span className="flex fr-icon-delete-line before:w-4 before:h-4 mr-2" />
          Supprimer
        </button>
      </div>
    )}
  >
    <OptionMenuButton />
  </DropdownFloater>
);

/** Bouton pour ouvrir le menu d'une option */
const OptionMenuButton = forwardRef((props, ref?: Ref<HTMLDivElement>) => (
  <div
    ref={ref}
    className="flex ml-6"
    onClick={evt => {
      evt.stopPropagation();
    }}
  >
    {/** Donne les props à un élément enfant afin de pouvoir donner le stopPropagation au onClick du parent */}
    <div
      {...props}
      className="my-auto p-1 rounded-lg cursor-pointer hover:bg-primary-2"
    >
      <span className="flex fr-icon-more-line before:w-5 before:h-5" />
    </div>
  </div>
));
OptionMenuButton.displayName = 'OptionOpenFloaterButton';
