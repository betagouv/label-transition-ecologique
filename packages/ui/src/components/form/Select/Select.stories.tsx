import React, {useState} from 'react';

import Select from './Select';
import {Meta, StoryObj} from '@storybook/react';

const options = [
  {value: 'option1', label: 'Option 1'},
  {value: 'option2', label: 'Option 2'},
  {
    value: 'option3',
    label:
      "Un titre d'option super long, car nous avons besoin de voir le comportement",
  },
];

const optionsWithSections = [
  {value: 'ss1', label: 'Option sans section 1'},
  {value: 'ss2', label: 'Option sans section 2'},
  {
    title: 'Section 1',
    options: [
      {value: 'option1', label: 'Option 1'},
      {value: 'option2', label: 'Option 2'},
      {
        value: 'option3',
        label:
          "Un titre d'option super long, car nous avons besoin de voir le comportement",
      },
    ],
  },
  {
    title: 'Section 2',
    options: [
      {value: 'option4', label: 'Option 4'},
      {value: 'option5', label: 'Option 5'},
    ],
  },
];

const meta: Meta<typeof Select> = {
  component: Select,
  decorators: [story => <div className="w-full max-w-[24rem]">{story()}</div>],
};

export default meta;

type Story = StoryObj<typeof Select>;

export const Default: Story = {
  args: {options},
  render: args => {
    const [value, setValue] = useState<string | undefined>(undefined);
    return (
      <Select
        {...args}
        values={value}
        onSelect={v => {
          setValue(v[0]);
        }}
      />
    );
  },
};

export const Disabled: Story = {
  args: {options, onSelect: () => null, disabled: true},
};

export const MultiSelectWithSectionOptions: Story = {
  args: {options: optionsWithSections, isMulti: true},
  render: args => {
    const [values, setValues] = useState<string[] | undefined>(undefined);
    return (
      <Select
        {...args}
        values={values}
        onSelect={v => {
          setValues(v);
        }}
      />
    );
  },
};

export const InputSelect: Story = {
  args: {options: optionsWithSections, isMulti: true},
  render: () => <div>TODO</div>,
};

// ne pas oublier le menu sur chaque option
export const CreateOptionSelect: Story = {
  args: {options: optionsWithSections, isMulti: true},
  render: () => <div>TODO</div>,
};
