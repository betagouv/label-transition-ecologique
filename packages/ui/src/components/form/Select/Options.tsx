import {Fragment} from 'react';
import classNames from 'classnames';

import {isOptionSection} from './utils';
import {OptionMenu} from './OptionMenu';
import {CreateOption} from './Select';

/**
 * Types partagés entre tous les composants selects
 * (Select, MultiSelect, MultiSelectFilter)
 */
export type OptionValue = number | string;
export type Option = {value: OptionValue; label: string};
export type OptionSection = {
  title: string;
  options: Option[];
};
export type SelectOption = Option | OptionSection;

type Props<T extends OptionValue> = {
  values?: T[];
  options: SelectOption[];
  onChange: (value: T) => void;
  isLoading: boolean;
  createProps?: CreateOption;
  renderOption?: (option: Option) => React.ReactElement;
  noOptionPlaceholder?: string;
  dataTest?: string;
};

/** Liste d'options pouvant être de simples options ou des sections */
const Options = <T extends OptionValue>({
  values,
  options,
  onChange,
  isLoading,
  renderOption,
  createProps,
  noOptionPlaceholder,
  dataTest,
}: Props<T>) => {
  return (
    <div data-test={`${dataTest}-options`}>
      {isLoading ? (
        <div className="p-4 text-sm text-gray-500">Chargement...</div>
      ) : options.length > 0 ? (
        options.map((option, i) => {
          /** Section */
          if (isOptionSection(option)) {
            return (
              <Fragment key={option.title + i}>
                <div className="first:hidden h-[1px] mx-6 bg-grey-4" />
                <div>
                  {/** titre */}
                  <div className="pt-4 pb-2 mx-6 font-bold text-left text-sm uppercase text-primary-7">
                    {option.title}
                  </div>
                  {/** options */}
                  {option.options.map(option => (
                    <Option
                      key={`${option.value}`}
                      option={option}
                      values={values}
                      onChange={onChange}
                      renderOption={renderOption}
                      createProps={createProps}
                    />
                  ))}
                </div>
              </Fragment>
            );
            /** Option simple */
          } else {
            return (
              <Option
                key={option.value}
                option={option}
                values={values}
                onChange={onChange}
                renderOption={renderOption}
                createProps={createProps}
              />
            );
          }
        })
      ) : (
        <div className="py-4 px-6 text-sm text-gray-500">
          {noOptionPlaceholder || 'Aucune option disponible'}
        </div>
      )}
    </div>
  );
};

export default Options;

type OptionProps<T extends OptionValue> = {
  values?: T[];
  option: Option;
  onChange: (value: T) => void;
  renderOption?: (option: Option) => React.ReactElement;
  createProps?: CreateOption;
};

/** Option pour les sélecteurs */
const Option = <T extends OptionValue>({
  values,
  option,
  onChange,
  renderOption,
  createProps,
}: OptionProps<T>) => {
  const isActive = values?.includes(option.value as T);
  return (
    <button
      data-test={option.value}
      className="flex items-start w-full p-2 pr-6 text-left text-sm hover:!bg-primary-0"
      onClick={() => onChange(option.value as T)}
    >
      <div className="flex w-6 mr-2 shrink-0">
        {isActive && (
          <span className="fr-icon-check-line flex mt-1 mx-auto text-primary-7 before:m-auto before:!h-4 before:!w-4" />
        )}
      </div>
      <div className="flex mr-auto">
        {renderOption ? (
          renderOption(option)
        ) : (
          <span
            className={classNames('leading-6 text-grey-8', {
              'text-primary-7': isActive,
            })}
          >
            {option.label}
          </span>
        )}
      </div>
      {createProps &&
        createProps.userCreatedOptions.some(o => o === option.value) && (
          <OptionMenu option={option} createProps={createProps} />
        )}
    </button>
  );
};
