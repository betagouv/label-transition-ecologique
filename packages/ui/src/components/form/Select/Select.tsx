import {Placement} from '@floating-ui/react';
import {Ref, forwardRef, useEffect, useRef, useState} from 'react';
import classNames from 'classnames';
import {useDebouncedCallback} from 'use-debounce';

import DropdownFloater from '../../floating-ui/DropdownFloater';
import Options, {OptionValue, SelectOption} from './Options';
import {Badge} from '../../badge/Badge';

import {
  filterOptions,
  getOptionLabel,
  getFlatOptions,
  sortOptionByAlphabet,
} from './utils';

export type CreateOption = {
  userCreatedOptions: OptionValue[];
  onCreate: (inputValue: string) => void;
  onDelete?: (id: OptionValue) => void;
  onUpdate?: (id: OptionValue, inputValue: string) => void;
};

type SelectProps<T extends OptionValue> = {
  /** Id pour les tests e2e */
  dataTest?: string;
  /** Liste des options */
  options: Array<SelectOption>;
  /** Appelée au click d'une option (reçoit la valeur de l'option cliquée) */
  onChange: (value: T) => void;
  /** Valeurs sélectionnées, peut recevoir une valeur seule ou un tableau de valeurs */
  values?: T | T[];
  /** Active la multi sélection */
  multiple?: boolean;
  /** Permet la recherche dans la liste d'option */
  isSearcheable?: boolean;
  /** Fonction exécutée lorsque l'utilisateur fait une recherche, reçoit la valeur de l'input */
  onSearch?: (v: string) => void;
  /** Temps du debounce appliqué à onSearch */
  debounce?: number;
  /** Les propriétés permettant la création de nouvelles options */
  createProps?: CreateOption;
  /** Fait apparaître un état de chargement à la place des options */
  isLoading?: boolean;
  /** Permet de désactiver le bouton d'ouverture */
  disabled?: boolean;
  /** Texte affiché quand rien n'est sélectionné */
  placeholder?: string;
  /** Texte affiché quand aucune option ne correspond à la recherche */
  emptySearchPlaceholder?: string;
  /** Change le positionnement du dropdown menu */
  placement?: Placement;
  /** Pour que la largeur des options ne dépasse pas la largeur du bouton d'ouverture */
  containerWidthMatchButton?: boolean;
};

/**
 * Composant multi-fonction:
 *
 * Select /
 * Multi select /
 * Searchable select /
 * Create option select
 *
 * Mettre `multiple` à `true` pour faire un multi select
 *
 * Ajouter `isSearcheable`, `createProps` ou `onSearch` pour faire un Searchable select
 *
 * Donner `createProps` pour faire un Create option select, cela active `isSearcheable' automatiquement
 */
export const Select = <T extends OptionValue>(props: SelectProps<T>) => {
  const {
    dataTest,
    values,
    options,
    onChange,
    createProps,
    onSearch,
    debounce = onSearch ? 250 : 0,
    placeholder,
    emptySearchPlaceholder,
    placement,
    multiple = false,
    isSearcheable = false,
    isLoading = false,
    containerWidthMatchButton = true,
    disabled = false,
  } = props;

  const hasSearch = isSearcheable || !!createProps || !!onSearch;

  /** Recherche textuelle locale car `onSearch` n'est pas obligatoire */
  const [inputValue, setInputValue] = useState('');

  /**
   * Permet de profiter du debounce de l'input et d'afficher un text de chargement
   * quand l'utilisateur est entrain de faire une saisie.
   */
  const [loading, setLoading] = useState(isLoading);
  // synchronise l'état de loading interne avec l'externe
  useEffect(() => setLoading(isLoading), [isLoading]);

  /** Fonction de debounce */
  const handleDebouncedInputChange = useDebouncedCallback(v => {
    onSearch(v);
    setLoading(false);
  }, debounce);

  /** Permet synchroniser les différents inputChange */
  const handleInputChange = (value: string) => {
    setInputValue(value);
    // uniquement si la fonction `onSearch` est donnée
    // on applique la fonction de `debounce`, par défaut le debounce est à 0
    if (onSearch) {
      // on active le loading sachant qu'on le désactive à la fin de la fct de debounce
      setLoading(true);
      handleDebouncedInputChange(value);
    }
  };

  /** Liste d'option filtrée par la saisie de l'utilisateur */
  const filteredOptions = filterOptions(options, inputValue);

  /** TODO: implémenter les action update et delete pour autoriser l'utilisation de ce cas */
  const isCreateOptionSelect = createProps !== undefined;
  // const isCreateOptionSelect = false;

  /** Compare la valeur de l'input de recherche avec la première optin de la liste
   * pour afficher le bouton de création d'une option */
  const isNotSimilar =
    inputValue.toLowerCase().trim() !==
    getFlatOptions(filteredOptions)[0]?.label.toLowerCase().trim();

  /** Transforme une valeur simple en tableau, qui est plus facile à traiter dans les sous composants */
  const arrayValues = values
    ? Array.isArray(values)
      ? values
      : [values]
    : undefined;

  return (
    <DropdownFloater
      placement={placement}
      offsetValue={0}
      containerWidthMatchButton={containerWidthMatchButton}
      noDropdownStyles
      multipleClickToggle={!hasSearch}
      enterToToggle={!hasSearch}
      disabled={disabled}
      render={({close}) => (
        <div
          data-test={`${dataTest}-options`}
          className="bg-white rounded-b-lg border border-grey-4 border-t-0 overflow-hidden"
        >
          {/** Bouton de création d'une option */}
          {isCreateOptionSelect &&
            inputValue.trim().length > 0 &&
            isNotSimilar && (
              <button
                data-test={`${dataTest}-create-option`}
                className="flex items-start justify-between w-full py-2 pl-10 pr-6 text-left text-sm hover:!bg-primary-0"
                onClick={() => {
                  createProps.onCreate(inputValue);
                  handleInputChange('');
                }}
              >
                <Badge title={inputValue} state="standard" />
                <span className="mt-1 ml-6 font-medium text-grey-8">Créer</span>
              </button>
            )}
          {/** Liste des options */}
          <Options
            dataTest={dataTest}
            values={arrayValues}
            options={
              isCreateOptionSelect
                ? sortOptionByAlphabet(filteredOptions)
                : filteredOptions
            }
            onChange={value => {
              onChange(value);
              setInputValue('');
              if (!multiple) {
                close();
              }
            }}
            isLoading={loading}
            createProps={createProps}
            noOptionPlaceholder={emptySearchPlaceholder}
          />
        </div>
      )}
    >
      <SelectButton
        dataTest={dataTest}
        values={arrayValues}
        options={options}
        onChange={onChange}
        isSearcheable={hasSearch}
        inputValue={inputValue}
        onSearch={handleInputChange}
        multiple={multiple}
        placeholder={placeholder}
        disabled={disabled}
      />
    </DropdownFloater>
  );
};

type SelectButtonProps<T extends OptionValue> = SelectProps<T> & {
  /** Donné par le DropdownFloater */
  isOpen?: boolean;
  /** Valeur de la saisie */
  inputValue: string;
  /** Change la valeur de saisie */
  onSearch: (value: string) => void;
};

/* Création d'un composant séparé pour passer la ref du boutton au floater */
const SelectButton = forwardRef(
  <T extends OptionValue>(
    {
      dataTest,
      isOpen,
      values,
      options,
      inputValue,
      isSearcheable,
      onSearch,
      multiple,
      placeholder,
      disabled,
      onChange,
      ...props
    }: SelectButtonProps<T>,
    ref?: Ref<HTMLButtonElement>
  ) => {
    const inputRef: Ref<HTMLInputElement> = useRef(null);

    /** donne le focus à l'input quand l'utilisateur click sur le bouton d'ouverture du Select */
    const handleWrapperClick = () => {
      inputRef?.current?.focus();
    };

    useEffect(() => {
      if (!isOpen) {
        onSearch('');
      }
    }, [isOpen]);

    return (
      <button
        ref={ref}
        data-test={dataTest}
        aria-expanded={isOpen}
        aria-label="ouvrir le menu"
        className={classNames(
          'w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden',
          {'rounded-b-none': isOpen}
        )}
        disabled={disabled}
        {...props}
      >
        {/**
         * Cette `div` englobant les sous composant permet d'appliquer la fonction pour forcer le focus sur l'input.
         * On ne peut pas l'appliquer sur le bouton car il reçoit déjà une propriété `onClick` du dropdown floater
         */}
        <div
          className="flex min-h-[3rem] py-2 px-4"
          onClick={handleWrapperClick}
        >
          <div className="flex grow flex-wrap gap-2 mr-4">
            {values && Array.isArray(values) ? (
              /** Listes des valeurs sélectionnées */
              <div className="flex items-center gap-2 grow">
                <Badge
                  title={getOptionLabel(values[0], getFlatOptions(options))}
                  onClose={!disabled && (() => onChange(values[0]))}
                />
                {/** Nombre de valeurs sélectionnées supplémentaires */}
                {values.length > 1 && (
                  <Badge title={`+${values.length - 1}`} state="info" />
                )}
              </div>
            ) : (
              /** Si pas de valeur et que la recherche n'est pas activée, on affiche un placeholder */
              !isSearcheable && (
                <span
                  className={classNames(
                    'my-auto text-xs text-grey-6 line-clamp-1',
                    {
                      '!text-grey-5': disabled,
                    }
                  )}
                >
                  {placeholder ?? multiple
                    ? 'Sélectionner une ou plusieurs options'
                    : 'Sélectionner une option'}
                </span>
              )
            )}
            {
              // on affiche l'input si le sélecteur est désactivé et ne possède pas de valeur
              // afin d'afficher le placeholder de l'input
              isSearcheable && !(disabled && values) && (
                <input
                  ref={inputRef}
                  type="text"
                  className="w-full text-sm outline-0 placeholder:text-grey-6"
                  value={inputValue}
                  onChange={e => onSearch(e.target.value)}
                  placeholder={placeholder ?? 'Rechercher par mots-clés'}
                  disabled={disabled}
                />
              )
            }
          </div>
          {/** Icône flèche d'ouverture */}
          <span
            className={classNames(
              'flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4',
              {'before:rotate-180': isOpen},
              {'before:text-grey-5': disabled}
            )}
          />
        </div>
      </button>
    );
  }
);
SelectButton.displayName = 'SelectButton';
