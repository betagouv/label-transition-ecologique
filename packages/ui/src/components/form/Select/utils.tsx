import {Option, OptionSection, SelectOption} from './Options';

/** Option section type guards */
export function isOptionSection(option: SelectOption): option is OptionSection {
  return (option as OptionSection).title !== undefined;
}

/** Option type guards */
export function isOption(option: SelectOption): option is Option {
  type NewType = Option;
  return (option as NewType).value !== undefined;
}

/** Renvoie un tableau d'options, quelles soient dans une section ou non */
export const getOptions = (selectOptions: SelectOption[]): Option[] =>
  selectOptions.reduce(
    (acc: Option[], curr) =>
      acc.concat(isOptionSection(curr) ? curr.options : curr),
    []
  );
