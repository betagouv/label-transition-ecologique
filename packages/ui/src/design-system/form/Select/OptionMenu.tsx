import {Ref, forwardRef} from 'react';
import DropdownFloater from '../../../components/floating-ui/DropdownFloater';

import {CreateOption} from './Select';
import {Option} from './Options';
import {Icon} from '../../icons/Icon';
import {Button} from '../../buttons/button/Button';

type Props = {
  option: Option;
  createProps: CreateOption;
};

export const OptionMenu = ({option, createProps}: Props) => (
  <DropdownFloater
    offsetValue={{mainAxis: 6}}
    placement="top-end"
    noDropdownStyles
    render={() => (
      <div
        className="flex flex-col divide-y divide-x-0 divide-solid divide-grey-3 rounded overflow-hidden bg-white border border-grey-3"
        onClick={evt => {
          evt.stopPropagation();
        }}
      >
        <button
          className="flex items-center w-full py-2 pr-4 pl-3 text-sm text-grey-8"
          onClick={() => createProps.onUpdate(option.value, 'TODO')}
        >
          <Icon icon="edit-line" size="sm" className="mr-2" />
          Ã‰diter
        </button>
        <button
          className="flex items-center w-full py-2 pr-4 pl-3 text-sm text-grey-8"
          onClick={() => createProps.onDelete(option.value)}
        >
          <Icon icon="delete-bin-6-line" size="sm" className="mr-2" />
          Supprimer
        </button>
      </div>
    )}
  >
    <OptionMenuButton />
  </DropdownFloater>
);

/** Bouton pour ouvrir le menu d'une option */
const OptionMenuButton = forwardRef((props, ref?: Ref<HTMLButtonElement>) => (
  <div className="flex px-3 py-1.5 my-auto group-hover:bg-primary-0 hover:!bg-white">
    <Button
      ref={ref}
      icon="more-line"
      size="xs"
      variant="white"
      className="!p-1 m-auto"
      {...props}
    />
  </div>
));
OptionMenuButton.displayName = 'OptionOpenFloaterButton';
