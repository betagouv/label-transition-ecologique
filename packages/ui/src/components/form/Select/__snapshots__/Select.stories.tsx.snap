// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/Select CreateOption smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <input type="text"
               class="w-full text-sm outline-0 placeholder:text-grey-6"
               placeholder="Rechercher par mots-clés"
               value
        >
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select Default smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <span class="my-auto text-xs text-grey-6 line-clamp-1">
          Sélectionner une option
        </span>
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select Disabled smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
          disabled
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <span class="my-auto text-xs text-grey-6 line-clamp-1 !text-grey-5">
          Sélectionner une option
        </span>
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4 before:text-grey-5">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select DisabledSearchableSelectWithValue smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
          disabled
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <div class="flex items-center gap-2 grow">
          <div data-test="Badge-undefined"
               class="bg-grey-3 flex items-center gap-1 max-w-max px-3 py-0.5 rounded"
          >
            <span class="text-grey-6 font-bold uppercase text-sm text-left line-clamp-1">
              Option 1
            </span>
          </div>
        </div>
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4 before:text-grey-5">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select MultiSelectWithSectionOptions smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <span class="my-auto text-xs text-grey-6 line-clamp-1">
          Sélectionner une ou plusieurs options
        </span>
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select SearchableMultiSelect smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <input type="text"
               class="w-full text-sm outline-0 placeholder:text-grey-6"
               placeholder="Rechercher par mots-clés"
               value
        >
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select SearchableSelectWithDebouncedApiCallOnTyping smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <button aria-expanded="false"
          aria-label="ouvrir le menu"
          class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
  >
    <div class="flex min-h-[3rem] py-2 px-4">
      <div class="flex grow flex-wrap gap-2 mr-4">
        <input type="text"
               class="w-full text-sm outline-0 placeholder:text-grey-6"
               placeholder="Rechercher par mots-clés"
               value
        >
      </div>
      <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4">
      </span>
    </div>
  </button>
</div>
`;

exports[`components/form/Select WithField smoke-test 1`] = `
<div class="w-full max-w-[24rem]">
  <div class="flex-grow flex flex-col gap-3">
    <label>
      <div class="font-medium text-grey-8">
        Description de l'action
      </div>
      <span class="block mt-2 text-xs text-grey-6">
        Texte description additionnel
      </span>
    </label>
    <button aria-expanded="false"
            aria-label="ouvrir le menu"
            class="w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden"
    >
      <div class="flex min-h-[3rem] py-2 px-4">
        <div class="flex grow flex-wrap gap-2 mr-4">
          <span class="my-auto text-xs text-grey-6 line-clamp-1">
            Sélectionner une option
          </span>
        </div>
        <span class="flex mt-2 ml-auto fr-icon-arrow-down-s-line text-primary-9 before:h-4 before:w-4">
        </span>
      </div>
    </button>
    <div class="flex items-center text-info-1">
      <span class="flex mr-1 before:h-4 before:w-4 fr-icon-info-fill">
      </span>
      <span class="text-xs">
        Message d’information
      </span>
    </div>
  </div>
</div>
`;
